Add support for s3 cloud storage

tanweeralii opened the pull request:
-> Compiled and run
-> `Upload` and `Delete` function are yet to be tested

tanweeralii committed:
Add support for s3 cloud storage

tanweeralii committed:
Add support for s3 cloud storage

hariprasanths commented on docs/examples.md:
````diff
@@ -117,6 +117,32 @@ $ synchly --config remote-sync
+* `credentials.json` should contain two keys shown in following structure:
+```
+{
+  aws_access_key_id : "",
````
property values must be double quoted.

hariprasanths commented on src/remoteSync/remoteSync.js:
```diff
@@ -36,6 +38,21 @@ const setupConfig = async (jobName, key, isDebug, filePath = undefined) => {
+            folders = folders.map((f) => {
+                return {name: f.Name};
+            });
+            let s3ConfigLoc = await s3Inquirer.askRemoteLoc(s3ConfigBuck.s3ParentBucketId, folders);
```
"s3ConfigBuck" object has no property named "s3ParentBucketId".
Can you please check this out. (I think it should be "s3ParentBucket").

hariprasanths commented on docs/examples.md:
```diff
@@ -117,6 +117,32 @@ $ synchly --config remote-sync
+? Enter the absolute path of the service account key file: /home/foobar/credentials.json
+✔ Authentication success
+? Choose the remote bucket in which backups will be stored: BucketBackup
+? Choose the remote folder in which backups will be stored: Backups
```
Change this line to the actual output. (Choose the remote folder in BucketBackup in which backups will be stored)
Can you please double quote the bucket name in docs as well as in actual program, as in - Choose the remote folder in "BucketBackup" in which backups will be stored

hariprasanths commented on src/remoteSync/s3/s3.js:
```diff
@@ -0,0 +1,84 @@
+
+let cloneServiceAccKey = async (jobName, key, serviceKeyLoc) => {
+    const jobConfStore = new configstore({configName: jobName, encryptionKey: key});
+    console.log(serviceKeyLoc);
```
Remove debug logs

hariprasanths commented on src/remoteSync/s3/s3.js:
```diff
@@ -0,0 +1,84 @@
+        Body: fileContent,
+    };
+    const res = await s3.upload(params).promise();
+    console.log(res);
```
Remove dubug logs

hariprasanths commented on src/remoteSync/s3/s3.js:
```diff
@@ -0,0 +1,84 @@
+
+    const params = {
+        Bucket: jobConfStore.get('s3ParentBucket'),
+        Key: jobConfStore.get('s3ParentFolder') + '/' + fileName,
```
Test the key part, I think the '/' will be considered a new folder.

hariprasanths commented on src/remoteSync/s3/s3.js:
```diff
@@ -0,0 +1,84 @@
+    let s3 = init(jobName, key);
+    const params = {
+        Bucket: jobConfStore.get('s3ParentBucket'),
+        Key: jobConfStore.get('s3ParentFolder') + '/' + fileName,
```
Test the key part, I think the '/' will be considered a new folder.

hariprasanths commented on src/remoteSync/remoteSync.js:
```diff
@@ -66,6 +83,8 @@ let uploadFile = async (jobName, key, fileName, filePath) => {
         resp = await gDrive.uploadFile(jobName, key, fileName, filePath);
     } else if (remoteType == 'SFTP') {
         resp = await sftp.uploadFile(jobName, key, fileName, filePath);
+    } else if (remoteType == 's3') {
```
remoteType should be compared with 'S3' instead of 's3'.

hariprasanths commented on src/remoteSync/remoteSync.js:
```diff
@@ -79,6 +98,8 @@ let deleteFile = async (jobName, key, fileName) => {
         resp = await gDrive.deleteFile(jobName, key, fileName);
     } else if (remoteType == 'SFTP') {
         resp = await sftp.deleteFile(jobName, key, fileName);
+    } else if (remoteType == 's3') {
```
remoteType should be compared with 'S3' instead of 's3'.

hariprasanths commented on src/remoteSync/s3/validator.js:
```diff
@@ -0,0 +1,27 @@
+const validateInitConfig = async (config) => {
+    let validatedConfig = {};
+
+    if (!config[s3ConfigKeys.s3AccKeyLoc]) {
```
Add docs for configuring using file:
https://github.com/hariprasanths/synchly/blob/master/docs/configuration-using-file.md#configuration-using-file

You can test it using this command: synchly --config=remote-sync --file=/home/foo/config.json

hariprasanths commented on src/remoteSync/s3/validator.js:
```diff
@@ -0,0 +1,27 @@
+const files = require('./../../utils/files');
+
+const s3ConfigKeys = {
+    s3AccKeyLoc: 's3sdkPath',
```
Change to user understandable name. ("s3sdkPath" to "s3CredentialsFilePath" maybe).

tanweeralii committed:
Fix `uploadFile` and `deleteFile` functions in `src/remoteSync/s3/s3.js`
Remove logs
Add example of config through files in `docs/configuration-using-file.md`
Fix other typo

tanweeralii committed:
Fix push error
